### PR TITLE
Preserve metadata changes for models (via editor) 

### PR DESCRIPTION
**Pull request by kulyk**
Note: this PR contains both FE and BE changes

This PR allows changing field metadata on a data model/dataset via the editor. The editor modifies queries `results_metadata` locally and once the "Save Changes" button is clicked, they're updated via `PUT /api/card/:id`

### To Verify

I really suggest watching the demo video below before testing the changes to get a better idea of how the editor works and which behaviors are worth testing

For both GUI and native queries:

1. Turn a question into a data model (prev. dataset)
2. Click the data model title to open the details sidebar
3. See the metadata completeness indicator next to the "Customize metadata" button, it should be below 100% (the way it works is described in detail in #19194)
4. Click "Customize metadata"
5. Try editing fields metadata (display name, description, semantic type, formatting, etc.)
6. Optional: try mixing editing the query and the metadata simultaneously
7. Click the "Save changes"  button at the top right at some point
8. Ensure the changes are persisted
9. Try creating a question based on the data model and make sure metadata overwrites work and remain visible (e.g. if you set a field to be an FK, the notebook editor should be able to automatically set a join condition by FK metadata)

**Extra flows**

1. Try to do a few data model edits (change metadata > Save > change metadata > Save > change smth else > Save, etc.) and try reverting to a particular revision, metadata changes should be kept
2. For GUI data models: play around with custom columns metadata editing. Create a few columns and see if their metadata is handled properly

### Demo

https://user-images.githubusercontent.com/17258145/149556601-f50d2b16-4b8d-4f6a-bcd0-c80705122007.mov

### CODE REVIEW

**Comment by nemanjaglumac** [comment]
Nit and something that doesn't block this PR, but should be addressed before we ship 42  - why is metadata sidebar "jumping" and animating every time you click on some other column header?

**Comment by nemanjaglumac** [comment]
What should happen in this scenario?

1. Open metadata editor
2. Edit column name "Address" to "Home Address"
3. Switch to the Query Editor
4. Add custom column
5. `concat([A` gives the `Address` column as the suggestion - it is not possible to have `Home Address`

**Review by nemanjaglumac — APPROVED** [review]
Apart from some comments I left, this seems to be working as described.
Giving this approval based on the local testing **only**. I didn't go through the code in detail.

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetFieldMetadataSidebar/MappedFieldPicker/MappedFieldPicker.tsx [hunk 1]
```diff
@@ -52,13 +89,22 @@ function MappedFieldPicker({ dataset, tabIndex }: MappedFieldPickerProps) {
         </StyledSelectButton>
       );
     },
-    [tabIndex],
+    [fieldObject, tabIndex],
   );
 
+  // TODO explain the need to skip virtual card id
```
Comment (daltojohnso) [inline frontend/src/metabase/query_builder/components/DatasetEditor/DatasetFieldMetadataSidebar/MappedFieldPicker/MappedFieldPicker.tsx:95]: leftover todo here

File: frontend/src/metabase/visualizations/components/TableInteractive.jsx [hunk 2]
```diff
@@ -546,11 +548,20 @@ export default class TableInteractive extends Component {
...
 
-  tableHeaderRenderer = ({ key, style, columnIndex }) => {
+  // TableInteractive renders invisible columns to remeasure the layout (see the _measure method)
+  // After the measurements are done, invisible columns get unmounted.
+  // Because table headers are wrapped into react-draggable, it can trigger
+  // https://github.com/react-grid-layout/react-draggable/issues/315
+  // (inputs loosing focus when draggable components unmount)
+  // We need to prevent that by passing `enableUserSelectHack={false}`
+  // to draggable components used in measurements
+  // We should maybe rethink the approach to measurements or render a very basic table header, without react-draggable
```
Comment (daltojohnso) [inline frontend/src/metabase/visualizations/components/TableInteractive.jsx:558]: +1 to this comment

File: frontend/src/metabase/query_builder/selectors.js [hunk 3]
```diff
@@ -47,9 +47,50 @@ export const getOriginalCard = state => state.qb.originalCard;
...
-  state.qb.queryResults && state.qb.queryResults[0];
+
+export const getMetadataDiff = state => state.qb.metadataDiff;
+
+const getRawQueryResults = state => state.qb.queryResults;
+
+export const getQueryResults = createSelector(
+  [getRawQueryResults, getMetadataDiff],
+  (queryResults, metadataDiff) => {
+    // return queryResults;
```
Comment (daltojohnso) [inline frontend/src/metabase/query_builder/selectors.js:58]: errant comment

**Review by daltojohnso — APPROVED** [review]
Played around with it and looked through the code -- looks good. I didn't look too closely at the clj.

I think the `DimensionInfoPopover`s don't receive metadata updates immediately because there's a `shouldComponentUpdate` method defined for the `TableInteractive` component and we'll need to update that.

I think in the next release or two there are a number of things we should start rethinking for the query builder. The "mode" it is in needs to be made more first class, and a hardening of metadata data structures to avoid all the `isEqual` and `isSame` checks we make would be nice.

**Review by mazameli — APPROVED** [review]
I think all the nits I noted in that video in Slack are separate concerns from the changes in this PR, so I'm approving. Nice work @kulyk and @dpsutton 🎉

File: src/metabase/api/dataset.clj [hunk 4]
```diff
@@ -55,10 +55,14 @@
       (events/publish-event! :table-read (assoc (Table table-id) :actor_id api/*current-user-id*))))
   ;; add sensible constraints for results limits on our query
   (let [source-card-id (query->source-card-id query)
-        info           {:executed-by api/*current-user-id*
-                        :context     context
-                        :card-id     source-card-id
-                        :nested?     (boolean source-card-id)}]
+        source-card    (when source-card-id
+                            (Card source-card-id))
```
Comment (camsaul) [inline src/metabase/api/dataset.clj:59]:
```suggestion
        source-card    (when source-card-id
                         (db/select-one [Card :result_metadata :dataset] :id source-card-id))
```


You're only using those two columns and it's extremely easy to limit the query to just those so might as well do it

**Review by camsaul — APPROVED** [review]
The backend stuff LGTM. The results metadata checksum has outlived its usefulness

**Pull request merged**